ruby: Enable all protocols in a single gem5 build

powerjg opened the pull request:
This PR includes the changes needed to enable all Ruby protocols in a single gem5 build.

Right now, this is just a simple rebase from the old changes on gerrit. See https://gem5-review.googlesource.com/c/public/gem5/+/59112 for details.

Things that are to do on this PR before reviewing (i.e., please ignore for now).
- [x] Implement a `Protocol` class (see comments here https://gem5-review.googlesource.com/c/public/gem5/+/59113/10)
- [x] Make sure it builds (have not tested since rebase)
- [x] Update stdlib changes
- [x] Update CHI changes
- [x] Test backwards compatible with single protocol
- [x] Test GPU FS (split) protocol
- [x] Test multi-protocol build 

To do as of 11/1
- [ ] Dig into problem with ViperBoard if rebased on this PR
- [ ] Change the `ALL` build option to build all protocols as well and update the tests to use this build

powerjg committed:
mem-ruby: Use shared and per-protocol SLICC files

This changes extends SLICC to understand two different kinds of slicc
files: files that are protocol-specific and files that are shared or
included between different protocols.

Each declaration in SLICC can now be shared or not. If it is shared,
then we can take a different action in the code generation (e.g., wrap
in a namespace).

*Developer facing change*
Removes the RubySlicc_interfaces.slicc file from the SLICC includes of
every protocol.

Changes required: If you have a custom protocol, you will need to remove
the line `include "RubySlicc_interfaces.slicc" from your .slicc file.

Change-Id: Ia6c2dafe2b8fe86749a13d17daa885bddd166855
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Use namespaces for protocol types

Wrap all protocol-specific types in `namespace <protocol>`. This will
facilitate compiling multiple protocols into one binary.

There is a one-time hack to the generated `MachineType.cc` file to use
the namespace for the protocol until we generalize the machine types.

Change-Id: I5947e8ac69afe6f7ed257d7c5980ad65e9338acf
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Move protocol files to subdir

Move all generated protocol-specific files to a subdirectory with the
protocol's name.

This change also updates SLICC to have separate variables for the
filename, c identifier and python identifier instead of just using
variations of the c identifier.

Change-Id: I62f69a4606b030ee23cb2d96493f3257a6923748
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Update HTML output to include protocol

Move the html output to be in a subdirectory with the protocol name.

Change-Id: I1510d2d5a531cc6db74d10a0478c23bc8a836a26
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs,mem-ruby: Protocol-spec. in learning gem5

Use protocol-specific names in Learning gem5 configs. Now, we should no
longer use the generic names for the controllers (it's deprecated). This
updates Learning gem5.

Website changes coming soon. (Hopefull before I push this...)

Change-Id: I18fc5b8bb0fef7c3b8b5cea8de4f73fc0f66a1b3
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
configs,mem-ruby: Procotol-spec. names in CHI

Use the protocol-specific controller names in CHI.

**Important**: This could change some scripts. As long as people use
CHI_config (likely), this shouldn't be a problem, but if you have a
different version of CHI_config.py locally, you will need to make the
following updates:

`Cache_Controller` -> `CHI_Cache_Controller`
`Memory_Controller` -> `CHI_Memory_Controller`

Website updates coming soon!

Change-Id: I7afdcede884ac5f9a9a76cc3d3dd35941e4e2faa
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
stdlib,mem-ruby: Use protocol-spec. names

Update the standard library Ruby protocols to use the protocol-specific
class names instead of the deprecated general names.

Unfortunately, some code became duplicated between similar controllers.
I tried multiple inheritance, but it didn't work out for me. I think the
correct solution is to move some of the shared code down into the
generated python. That's out of the scope for these changes.

Change-Id: I3444bee3c2917dcbe92b600b85e60244129aad35
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
python: Expand Enum param type to be more Enum-like

This extends gem5's version of python enums to support an equal operator
and the hash operator so we can compare two instances of enums and add
these to sets/dicts/etc.

Change-Id: I4a785bf9570a54254ada1db684379ee77e67b192
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
tests: Add more hierarchies to traffic gen tests

This adds all of the hierarchies supported in the standard library. We
can soon move to using a different build target and run all hierarchies!

Change-Id: Ic065a679ea34c3bb2f71b3b133806d240039fbb5
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Rename SLICC SimObjs with compatibility

Rename all SLICC generated SimObjects to have the protocol in their
name. This will allow for two different protocols to have the same
machine names (e.g., L1Cache). For compatiblity, we check to see if the
current or main protocol that is built matches the SimObject's protocol
and export the backwards-compatible name.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Update MachineType autogen file with all types

This change makes it so that the MachineType.cc/hh file are not unique
for each protocol. All of the machine types are now tracked.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
scons: Change scons for multiple protocols in SLICC

This change is a step toward multiple protocols building at the same
time in scons. Add functions and use lists instead of single protocol.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby,configs: Update AMD protos with new names

Update the MOESI_AMD_Base and GPU_VIPER configuration files with the new
full protocol-specific names for the controllers instead of the
deprecated names.

Note: If you have any files which use the `CntrlBase` base, you will
likely need to update the class names that you are inheriting from.

Change-Id: I623fea7dd4cd151f7b15fe7cb43f8a4c45492d89
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Add ProtocolInfo class

Add a ProtocolInfo class that is specialized (through inheritance) for
each protocol. This class currently has the protocol's name and any
protocol-specific options (partial_func_reads is the only one so far).
Note that the SLICC language has been updated so that you can specify
the options in the `protocol` statement in the `.slicc` file.

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby: Use runtime protocol instead of #defines

This removes two #defines: PARTIAL_FUNC_READS and PROTOCOL_<protocol>.
Instead, update the code to use the runtime information about which
protocol we are using.

Change-Id: Icb6f10fc2d3fd59128c62f9f6e37b52ef2581b61
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
mem-ruby,scons: Add scons option for multiple protocols

This change does many things, but they must all be atomically done.

**USER FACING CHANGE**: The Ruby protocols in Kconfig have changed names
(they are now the same case as the SLICC file names). So, after this
commit, your build configurations need to be updated. You can do so by
running `scons menuconfig <build dir>` and selecting the right ruby
options. Alternatively, if you're using a `build_opts` file, you can run
`scons defconfig build/<ISA> build_opts/<ISA>` which should update your
config correctly.

Detailed changes are described below.

Kconfig changes:

- Kconfig files in ruby now must all be declared in the ruby/Kconfig
  file
- All of the protocol names are changed to match their slicc file names
  including the case
- A new option is available called "Use multiple protocols" which should
  be selected if multiple protocols are selected. This is only used to
  set the PROTOCOL variable to "MULTIPLE" when in multiple mode.
- The PROTOCOL variable can now be "MULTIPLE" which means it will be
  ignored. If it's not "MULTIPLE" then it holds the "main" protocol,
  which is necessary for backwards compatibility with the Ruby.py files.

Ruby config changes:

To make this change backwards compatible with Ruby.py, this change adds
a new "protocol" config called MULTIPLE.py which is used to allow the
user to set a "--protocol" option on the command line. This is only
needed if you are using a gem5 binary with multiple protocols but need
to use Ruby.py.

stdlib changes:

- Make the coherence protocol file behave like the ISA file
- Add a function to get the coherence protocol from the `CacheHierarchy`
  like we do with the ISA in the `Processor`.
  - Use this function where `get_runtime_coherence_protocol` was used
- Update the requires code to work with the ne CoherenceProtocol
- Fix a typo in the AMD Hammer name and also add the missing MSI
  protocol

Scons changes:

- In Ruby we now gather up all of the protocols and build them all if
  there are multiple protocols
- There's some bending over backwards to tell the user if they are using
  an out of date gem5.build/config file and how to update it
- Note that multiple ruby protocols adds a significant amount of time to
  the build since we have to run slicc twice for each file.

build_opts:

- Update all files with new names
- Add a new NULL_All_Ruby that will be used for testing

Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg commented:
🥳 This is now ready for review! There's still some more testing that needs to be done, and there are likely mistakes, but it's in a state that can be reviewed now.

Let me state one thing, for the record: I spent more time fighting with Kconfig than I did updating Ruby or anything else. This is not a good system. It is not flexible in the way we need, and it's not well documented.

Also, I've done my best to ensure backwards compatibility, but I could not get the build to work perfectly. See the RELEASE-NOTES.md for details.

I hope that this addresses most of the prior reviews.

For incoming reviews, let's try to concentrate on giving feedback if there are any bugs or corner cases I may have missed. Unfortunately, there is no chance that I'll have another 15+ hours to go back and redesign this change for the third time. This is the best design I could come up with, given the prior feedback and my limited coding cycles.

powerjg commented:
@BobbyRBruce the tests are failing because of caching. Is there a way to clear the build cache in the runners? Alternatively, if you (or anyone else) can figure out how to enable full backwards compatibility in the Kconfig, that would fix it. [This commit](https://github.com/gem5/gem5/pull/117/commits/7d4a7cd3eabba9be521a1f725d6cd2fa5c1a23f9) is the major problem.

BobbyRBruce commented:
> @BobbyRBruce the tests are failing because of caching. Is there a way to clear the build cache in the runners? Alternatively, if you (or anyone else) can figure out how to enable full backwards compatibility in the Kconfig, that would fix it. [This commit](https://github.com/gem5/gem5/pull/117/commits/7d4a7cd3eabba9be521a1f725d6cd2fa5c1a23f9) is the major problem.

I just wiped the build caches. For future reference you can delete any cache here: https://github.com/gem5/gem5/actions/caches

BobbyRBruce commented:
Failure to build VEGA_X86 is a bug with tnhe PR. It still happened when the cache was removed.

powerjg commented:
> Failure to build VEGA_X86 is a bug with tnhe PR. It still happened when the cache was removed.

I didn't say it was failing *only* because of caching ;)

BobbyRBruce commented on configs/ruby/MULTIPLE.py:
```diff
@@ -0,0 +1,89 @@
+# Copyright (c) Jason Lowe-Power
```
I'm not sure if you (a UC employee are allowed to claim this copyright 😀.

Regardless, this needs a year (I think... i don't claim to understand anything about copyright law).

powerjg replied:
I'll add years. As a professor, I (can) own copyright. UC is weird. See https://copyright.universityofcalifornia.edu/ownership/works-created-at-uc.html

BobbyRBruce reviewed (commented):
I have looked over this it broadly looks fine to me, though I think it's too complicated for me to approve just through code review review. Given how big this change is I would prefer if, in addition to the CI tests, the daily and weekly tests were run as well prior to merging.

powerjg commented:
> the daily and weekly tests were run as well prior to merging.

This is a good idea. @erin-le can you do this?

powerjg commented:
@abmerop not sure if you were notified of this change since I marked it as "ready for review"
cc: @giactra

abmerop commented on src/mem/ruby/Kconfig:
```diff
@@ -28,15 +28,28 @@ menu "Ruby"
+
+        config USE_MULTIPLE_PROTOCOLS
+            bool 'Use multiple protocols'
+            default y
```
Is this overridden by build_opts? I don't quite understand the interaction between build_opts / Kconfig.

If not, do we want to make default n until multiple protocols is better tested?

powerjg replied:
I'm sure there's a better way to do this. However, I spent literally half of my development time bringing this up to date on Kconfig changes, and I couldn't find a better way. Here's what I know.

- The files in `build_opts` provides defaults for the Kconfig config file.
- `USE_MULTIPLE_PROTOCOLS` is actually ignored completely in the code.
- I only use this variable to set the `PROTOCOL` variable to "MULTIPLE" by default.
- In the `build_opts`, files for backwards compatibility, each file specifies the name of the protocol in `PROTOCOL` which is the only thing that really matters.

And finally, the last thing I know is that I *do not* like Kconfig

abmerop commented on src/mem/ruby/Kconfig:
```diff
@@ -28,15 +28,28 @@ menu "Ruby"
-        config NEED_PARTIAL_FUNC_READS
-            bool
+        rsource "protocol/Kconfig"
+        rsource "../../learning_gem5/part3/Kconfig"
```
Was this intended to be moved from src/Kconfig to here?

powerjg replied:
Yes. This is another thing that caused me headaches. When moving the a "choice" to a "config" you have to specify all of the options in a single place. So, I had to move this here. I don't like it, but I also just don't like Kconfig.

abmerop commented on src/mem/ruby/system/RubySystem.cc:
```diff
@@ -493,9 +504,17 @@ RubySystem::resetStats()
     ClockedObject::resetStats();
 }
 
-#ifndef PARTIAL_FUNC_READS
```
Nice, this was always a source of confusion which on is being used at runtime. Should be much more clear now!

abmerop commented on src/mem/slicc/symbols/StateMachine.py:
```diff
@@ -302,38 +304,75 @@ class $py_ident(RubyController):
         code.dedent()
-        code.write(path, f"{py_ident}.py")
+
+        # Needed for backwards compatibility while there is only one protocol
+        # compiled. When moving to multiple protocols in the gem5 binary, this
+        # will need to change.
```
Hmm, is this comment still correct?  Does it imply that currently the PR doesn't currently support multiple protocols?

powerjg replied:
I'll update this comment. The code is correct, but the comment is out of date.

powerjg replied:
For rebasing reasons, this comment isn't going to be updated in the right commit, but the new text will be right after the merge.

abmerop commented on src/mem/ruby/protocol/SConscript:
```diff
@@ -34,87 +34,141 @@ from SCons.Scanner import Classic
-if not protocol_dir:
-    raise ValueError("Could not find {}.slicc in PROTOCOL_DIRS".format(
-        protocol))
+def slicc_emitter(target, source, env):
```
I didn't go through the next 80 lines *super* carefully since there is a lot I don't understand here, but if it has been tested with/without multiple protocols that's good enough for me to prove it works.

powerjg replied:
Yeah, it's been tested, though it can always be tested more!

One thing to note here is that we have to run slicc on all protocols *twice*. The first time (the emitter code code here, IIRC) we just figure out what files we will be making (which, annoyingly requires an entire slicc parsing). Then, in the action code below we actually make the files. Unfortunately, this causes gem5's build to slow down significantly in the early stages (and it's not parallelizable at all).

abmerop reviewed (commented):
This looks good to me!  I like the approach, how the files are organized, namespacing, etc. and it looks like this required a lot of knowledge of SLICC, Kconfig, SCons, python, *and* C++. So nothing major from me, but a few questions/comments.

powerjg commented:
@abmerop BTW, your code to remove the static variables really helped make the rebase easier! Thanks for your work on #1349!

abmerop commented:
I was able to test this PR with the current GPUFS scripts and they are working for square (one of the todos).

However, I updated the viper board after applying this PR and ran into a issue with RubyNetwork. Just wondering if you have seen this? `src/mem/ruby/network/simple/routing/WeightBased.cc:127: panic: assert(msg_dsts.count() == 0) failed`

erin-le committed:
tests: additional trusted_stats files

This commit adds trusted_stats.json files for CHIL1,
MESI_Three_Level, MIExample, and OctopiCache. These jsons are
used to verify the output of tests from test_memory_traffic_gen.py.

erin-le committed:
scons, ext-testlib: rename NULL_All_Ruby to use all caps

This commit renames NULL_All_Ruby to NULL_ALL_RUBY and adds a
tag for this build for testlib. These changes were made so the
NULL_All_Ruby build could be used with testlib.

erin-le committed:
mem-ruby: changes to MESIThreeLevel, MIExample, OctopiCache

This commit changes MESIThreeLevel, MIExample, and OctopiCache
so they work with this PR. It also adds MESIThreeLevel and
OctopiCache to the testlib tests.

erin-le commented:
I tested this using test_memory_traffic_gen.py and simple_traffic_run.py from the TestLib and had to make some modifications to get the tests to pass.

TODO: Still need to change the ALL build option to use all protocols and change the TestLib tests to use the new ALL build option.

powerjg commented:
> However, I updated the viper board after applying this PR and ran into a issue with RubyNetwork. Just wondering if you have seen this? `src/mem/ruby/network/simple/routing/WeightBased.cc:127: panic: assert(msg_dsts.count() == 0) failed`

Hmm, no I haven't seen that. But I also haven't tried to rebase the ViperBoard PR on top of this. I wonder if it's related to the disjoint network somehow.

abmerop commented:
> > However, I updated the viper board after applying this PR and ran into a issue with RubyNetwork. Just wondering if you have seen this? `src/mem/ruby/network/simple/routing/WeightBased.cc:127: panic: assert(msg_dsts.count() == 0) failed`
> 
> Hmm, no I haven't seen that. But I also haven't tried to rebase the ViperBoard PR on top of this. I wonder if it's related to the disjoint network somehow.

ViperBoard is not using using multiple networks, or at least it shouldn't be. Are you planning to try and rebase it?  If not I can try and take a look this weekend

powerjg commented:
No, I doubt I'll have time to rebase the ViperBoard on this change. I'll be traveling at MICRO through Wednesday and then out of the office Thurs-Friday.

I'll do my best to get the DArchR changes to the ViperBoard ready to go by Monday (depending on when I finish working on my talk and if I have extra time at workshops over the weekend).

abmerop commented:
> No, I doubt I'll have time to rebase the ViperBoard on this change. I'll be traveling at MICRO through Wednesday and then out of the office Thurs-Friday.
> 
> I'll do my best to get the DArchR changes to the ViperBoard ready to go by Monday (depending on when I finish working on my talk and if I have extra time at workshops over the weekend).

Sounds good. On Monday I can apply additional changes to the board including rebasing if this PR goes first and the UCD ViperBoard changes get push on top of my PR. That way I think both PRs can get in to v24.1.

I have found the issue I was describing above and it is specific to VIPER and not a problem with this PR. You can probably spot the problem here: https://github.com/abmerop/gem5/blob/viper-board/src/python/gem5/prebuilt/viper/viper_network.py#L94-L95

erin-le committed:
tests: modify tests to use ALL build

This commit modifies a number of Testlib tests to use the ALL
build instead of a more specific build.

erin-le committed:
scons: modify ALL build to have all Ruby protocols

erin-le committed:
tests, scons: Fix Testlib test failures

This commit changes the fs/linux/arm and learning_gem5 tests as
they were previously failing with the Ruby change. The
fs/linux/arm long tests require the addition of a new gem5 build,
ARM_X86, which builds the ARM and X86 ISAs with the
MESI_Two_Level cache hierarchy.

erin-le commented:
I pushed the changes to add all Ruby protocols to the ALL build and update the tests to use this build where possible. I wasn't able to get the `fs/linux/arm` Testlib tests to work with any of the existing gem5 builds, so I added an ARM_X86 build opt.

BobbyRBruce committed:
Merge branch 'develop' into new-ruby

BobbyRBruce commented on configs/learning_gem5/part3/msi_caches.py:
```diff
@@ -46,7 +46,7 @@
 class MyCacheSystem(RubySystem):
     def __init__(self):
-        if buildEnv["PROTOCOL"] != "MSI":
+        if not "RUBY_PROTOCOL_MSI" in buildEnv:
```
so it's `in buildEnv` and not `in buildEnv["PROTOCOL"]`? It's just searching for the key to know if the protocol is there? If so, what's the value? I.e., what do I get if I go with `print(buildEnv["RUBY_PROTOCOL_MSI"])`?

BobbyRBruce reviewed (dismissed):
I want to merge this PR soon. It seems to have had ample time for feedback and it needs merged to unblock other work (inc. the v24.1 release). If you have any protests on this please speak up ASAP!


I can't see any issues here. I think this is a great contribution.

abmerop reviewed (dismissed):
My comments were addressed and it's rebased to looks good to me.

FYI, the ViperBoard issue in the PR description has also been resolved, but is still marked as a TODO.

erin-le committed:
tests: remove protocol=None, add print statement back in

This commit removes the `protocol=None` argument in various
gem5_verify_config()s because protocol is set to None by
default. Also, two print statements that were taken out in
previous commits were put back in with different function calls.

BobbyRBruce committed:
arch-arm,misc: Fixing arch-arm SConscript

This was caused by a bad merge.

BobbyRBruce committed:
misc: Include build_opts updates in RELEASE_NOTES

powerjg commented on RELEASE-NOTES.md:
```diff
@@ -7,6 +7,19 @@
+  * **USER FACING CHANGE**: The Ruby protocols in Kconfig have changed names (they are now the same case as the SLICC file names), and in addition,  So, after this commit, your build configurations need to be updated. You can do so by running `scons menuconfig <build dir>` and selecting the right ruby options. Alternatively, if you're using a `build_opts` file, you can run `scons defconfig build/ ...
+  * **USER FACING CHANGE**: The the "build_opts/ALL" build spec has been updated to include all Ruby protocols . As such, gem5 compilations of the "ALL" compilation target will include all gem5 Ruby protocols (previously just MESI_Two_Level).
+  * A "build_opts/NULL_ALL_RUBY" build spec has been added to include all Ruby protocols for a "NULL ISA" build . This is useful for testing Ruby protocols without the overhead of a full ISA and is used in gem5's traffic generator tests.
+  * A "build_opts/ARM_X96" build spec has been added due to a unique restriction in the "tests/gem5/fs/linux/arm" tests which requires a compilation of gem5 with both ARM and X86 and solely the MESI_Two_Level protocol.
```
Typo X96 -> X86

BobbyRBruce replied:
I'll push a separate change after merging this :P

powerjg reviewed (commented):
Thanks for the help pushing this over the line! Some of the tests randomly failed, so I re-ran the failed tests. Hopefully we'll be good to go soon.

Maybe we can fix my comment when we update the release notes later in the cycle to prevent another round of CI tests?

BobbyRBruce commented:
> Thanks for the help pushing this over the line! Some of the tests randomly failed, so I re-ran the failed tests. Hopefully we'll be good to go soon.
> 
> Maybe we can fix my comment when we update the release notes later in the cycle to prevent another round of CI tests?

This was just due to too many runners trying to download too much stuff at the same time... it's a problem with our runners. We can have lots running at once but if they all start at roughly the same time they tend to all need to download lots of resources at the same time. The network grinds to a halt and the downloads start to timeout.

Anyway, I'm almost certain these tests will pass.

powerjg commented:
🥳 Our oldest PR gets committed!!!
